mypy/checker.py

```diff
@@ -7581,6 +7577,20 @@ def warn_deprecated(self, node: SymbolNode | None, context: Context) -> None:
             warn = self.msg.fail if self.options.report_deprecated_as_error else self.msg.note
             warn(deprecated, context, code=codes.DEPRECATED)
 
+    def search_deprecated(
+        self, typ: Type | None, s: AssignmentStmt, visited: set[Type | None]
+    ) -> None:
+
+        if typ not in visited:
+            visited.add(typ)
+            if isinstance(typ := get_proper_type(typ), Instance):
+                self.check_deprecated(typ.type, s)
+                for arg in typ.args:
+                    self.search_deprecated(arg, s, visited)
+            elif isinstance(typ, (UnionType, TupleType)):
```

Comment:
This is not a robust way to do it. You should write a simple visitor instead.

Reply:
I replaced `search_deprecated` with `InstanceDeprecatedVisitor`, which additionally considers `CallableType` and `TypeAliasType`.  Do you see other relevant cases?